server/services/client.ts

```diff
@@ -333,13 +333,37 @@ const clientService: (context: StrapiContext) => IClientService = ({ strapi }) =
         default:
           const publishedItems = items.filter(filterOutUnpublished);
           const result = isNil(rootPath) ? items : filterByPath(publishedItems, rootPath).items;
+
+          const defaultCache = new Map<Id, number[]>();
+          const getNestedOrders = (id: Id, cache: Map<Id, number[]> = defaultCache): number[] => {
+            const cached = cache.get(id);
+            if (!isNil(cached))
+              return cached;
+            
+            const item = result.find(item => item.id === id);
+            if (isNil(item))
+              throw new Error("Item not found");
+            
+            const { order, parent } = item;
+            if (parent) {
+              const nestedOrders = getNestedOrders(parent.id, cache).concat(order);
+              cache.set(id, nestedOrders);
+              return nestedOrders;
+            }
+            else {
+              const nestedOrders = [order];
+              cache.set(id, nestedOrders);
+              return nestedOrders;
+            }
+          }
+
           return result.map((item: NavigationItemEntity<ContentTypeEntity>) => ({
             ...item,
             audience: item.audience?.map(_ => (_).key),
             title: composeItemTitle(item, contentTypesNameFields, contentTypes) || '',
             related: wrapContentType(item.related),//omit(item.related, 'localizations'),
             items: null,
-          }));
+          })).sort((a, b) => compareArraysOfNumbers(getNestedOrders(a.id), getNestedOrders(b.id)));
```

Comment:
[minor]

IMO it would be a bit more readable if we had

```tsx
return result
  .map()
  .sort()
```

Reply:
I agree 👌